### ISSUE TITLE
model not compiling:  `object 'tf' not found`

### ISSUE THREAD

**Issue by bogsnork** [bogsnork/nectarvision#7]
https://github.com/bogsnork/nectarvision/blob/bc2adb6c9df66794381a575ce533fdda478f004d/R/ssd_training.R#L387   split from #5 .    Error: 

` Error in py_call_impl(callable, dots$args, dots$keywords) : 
  RuntimeError: Evaluation error: object 'tf' not found. `

**Comment by bogsnork** [comment]
Traceback: 

`21. | stop(structure(list(message = "RuntimeError: Evaluation error: object 'tf' not found.",      call = py_call_impl(callable, dots$args, dots$keywords),      cppstack = structure(list(file = "", line = -1L, stack = c("/usr/local/lib/R/site-library/reticulate/libs/reticulate.so(Rcpp::exception::exception(char const*, bool)+0x84) [0x7fea774ab2d4]",      "/usr/local/lib/R/site-library/reticulate/libs/reticulate.so(Rcpp::stop(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x27) [0x7fea774ab427]",  ...
`

**Comment by bogsnork** [comment]
R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

Matrix products: default
BLAS/LAPACK: /usr/lib/libopenblasp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] listviewer_2.1.0  repurrrsive_0.1.0 magick_2.0        rjson_0.2.20      keras_2.2.4.1    
 [6] forcats_0.4.0     stringr_1.4.0     dplyr_0.8.3       purrr_0.3.2       readr_1.3.1      
[11] tidyr_0.8.3       tibble_2.1.3      ggplot2_3.2.0     tidyverse_1.2.1  

loaded via a namespace (and not attached):
 [1] reticulate_1.12   tidyselect_0.2.5  haven_2.1.1       lattice_0.20-38   vctrs_0.1.0      
 [6] colorspace_1.4-1  generics_0.0.2    htmltools_0.3.6   base64enc_0.1-3   utf8_1.1.4       
[11] rlang_0.4.0       pillar_1.4.2      glue_1.3.1        withr_2.1.2       modelr_0.1.4     
[16] readxl_1.3.1      tensorflow_1.13.1 munsell_0.5.0     gtable_0.3.0      cellranger_1.1.0 
[21] rvest_0.3.4       htmlwidgets_1.3   labeling_0.3      tfruns_1.4        fansi_0.4.0      
[26] broom_0.5.2       Rcpp_1.0.1        scales_1.0.0      backports_1.1.4   jsonlite_1.6     
[31] digest_0.6.20     hms_0.4.2         packrat_0.5.0     stringi_1.4.3     grid_3.6.0       
[36] cli_1.1.0         tools_3.6.0       magrittr_1.5      lazyeval_0.2.2    crayon_1.3.4     
[41] whisker_0.3-2     pkgconfig_2.0.2   zeallot_0.1.0     Matrix_1.2-17     xml2_1.2.0       
[46] lubridate_1.7.4   assertthat_0.2.1  httr_1.4.0        rstudioapi_0.10   R6_2.4.0         
[51] nlme_3.1-139      compiler_3.6.0

**Comment by bogsnork** [comment]
CLICOLOR_FORCE               1
CUDA_HOME                    /usr/local/cuda
CUDA_PATH                    /usr/local/cuda
DISPLAY                      :0
EDITOR                       vi
GIT_ASKPASS                  rpostback-askpass
HOME                         /home/rstudio
LANG                         en_US.UTF-8
LD_LIBRARY_PATH              /usr/local/lib/R/lib:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64::/lib:/usr/local/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
LN_S                         ln -s
LOGNAME                      rstudio
MAGICK_TMPDIR                /tmp/RtmpTl2bpG
MAKE                         make
PAGER                        /usr/bin/pager
PATH                         /opt/virtualenvs/r-tensorflow/bin:/usr/local/cuda/bin:/usr/local/nvidia/bin:/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
R_BROWSER                    xdg-open
R_BZIPCMD                    /bin/bzip2
R_DOC_DIR                    /usr/local/lib/R/doc
R_GZIPCMD                    /bin/gzip
R_HOME                       /usr/local/lib/R
R_INCLUDE_DIR                /usr/local/lib/R/include
R_LIBS                       /usr/local/lib/R/site-library:/usr/local/lib/R/library:/usr/lib/R/library
R_LIBS_SITE                  
R_LIBS_USER                  /usr/local/lib/R/site-library
R_PACKRAT_DEFAULT_LIBPATHS   /usr/local/lib/R/site-library:/usr/local/lib/R/library
R_PACKRAT_SITE_LIBRARY       /usr/local/lib/R/site-library
R_PACKRAT_SYSTEM_LIBRARY     /usr/local/lib/R/library
R_PAPERSIZE                  letter
R_PDFVIEWER                  /usr/bin/xdg-open
R_PLATFORM                   x86_64-pc-linux-gnu
R_PRINTCMD                   /usr/bin/lpr
R_RD4PDF                     times,inconsolata,hyper
R_SESSION_INITIALIZED        PID=2278:NAME="reticulate"
R_SESSION_TMPDIR             /tmp/RtmpTl2bpG
R_SHARE_DIR                  /usr/local/lib/R/share
R_STRIP_SHARED_LIB           strip --strip-unneeded
R_STRIP_STATIC_LIB           strip --strip-debug
R_SYSTEM_ABI                 linux,gcc,gxx,gfortran,gfortran
R_TEXI2DVICMD                /usr/bin/texi2dvi
R_UNZIPCMD                   /usr/bin/unzip
R_ZIPCMD                     /usr/bin/zip
RETICULATE_PYTHON_ENV        /opt/virtualenvs/r-tensorflow
RETICULATE_REQUIRED_MODULE   tensorflow
RMARKDOWN_MATHJAX_PATH       /usr/lib/rstudio-server/resources/mathjax-26
RS_RPOSTBACK_PATH            /usr/lib/rstudio-server/bin/rpostback
RSTUDIO                      1
RSTUDIO_CONSOLE_COLOR        256
RSTUDIO_CONSOLE_WIDTH        98
RSTUDIO_PANDOC               /usr/lib/rstudio-server/bin/pandoc
RSTUDIO_R_MODULE             
RSTUDIO_R_PRELAUNCH_SCRIPT   
RSTUDIO_R_VERSION_LABEL      
RSTUDIO_SESSION_STREAM       rstudio-d
RSTUDIO_USER_IDENTITY        rstudio
RSTUDIO_VERSION              1.2.1335
RSTUDIO_WINUTILS             bin/winutils
SED                          /bin/sed
SSH_ASKPASS                  rpostback-askpass
TAR                          /bin/tar
TERM                         xterm-256color
USER                         rstudio
VIRTUAL_ENV                  /opt/virtualenvs/r-tensorflow
WORKON_HOME                  /opt/virtualenvs

**Comment by bogsnork** [comment]
Also tested install.tensorflow()

`Traceback (most recent call last):
  File "/opt/virtualenvs/r-tensorflow/lib/python3.5/site-packages/virtualenv.py", line 2610, in <module>
    main()
  File "/opt/virtualenvs/r-tensorflow/lib/python3.5/site-packages/virtualenv.py", line 834, in main
    symlink=options.symlink,
  File "/opt/virtualenvs/r-tensorflow/lib/python3.5/site-packages/virtualenv.py", line 1112, in create_environment
    install_python(home_dir, lib_dir, inc_dir, bin_dir, site_packages=site_packages, clear=clear, symlink=symlink)
  File "/opt/virtualenvs/r-tensorflow/lib/python3.5/site-packages/virtualenv.py", line 1385, in install_python
    mkdir(lib_dir)
  File "/opt/virtualenvs/r-tensorflow/lib/python3.5/site-packages/virtualenv.py", line 394, in mkdir
    os.makedirs(at_path)
  File "/usr/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/opt/virtualenvs/r-tensorflow/lib/python2.7'
Running virtualenv with interpreter /usr/bin/python
Error: Error 1 occurred creating virtualenv at /opt/virtualenvs/r-tensorflow`

**Comment by bogsnork** [comment]
This might help: https://github.com/rocker-org/ml/tree/master/ubuntu

> Managing the Python Virtualenv
> Currently, we set global env var PYTHON_RETICULATE_ENV, which forces reticulate and friends to use the virtualenv we set up at that location (opt/venv). This may make it dificult to provide alternative venvs (since running Sys.unsetenv() from the R console is not sufficient).
> 
> Setting WORKON_HOME is more flexible, e.g.:
> 
> ENV WORKON_HOME /opt/virtualenvs
> ENV PYTHON_VENV_PATH $WORKON_HOME/r-tensorflow
> which allows reticulate to still find python in the venv out-of-the-box, but can create additional virtualenvs in WORKON_HOME (users should be given write permissions to this).

**Comment by bogsnork** [comment]
tried to give user permission to /opt/virtualenvs by ssh into the instance and running
`sudo nvidia-docker exec rstudio_ml_1 chmod -R 777 /opt/virtualenvs`

**Comment by bogsnork** [comment]
> tried to give user permission to /opt/virtualenvs by ssh into the instance and running
> `sudo nvidia-docker exec rstudio_ml_1 chmod -R 777 /opt/virtualenvs`

didn't fix the error.

**Comment by bogsnork** [comment]
changed docker image to rocker/ml::cuda10.0
tried 
````r
install.tensorflow()
````
error: 
```
ERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/venv/lib/python3.6/site-packages/six.py'
Consider using the `--user` option or check the permissions.

Error: Error installing package(s): 'tensorflow==2.0.0'
```

**Comment by bogsnork** [comment]
In this build, opt/virtualenvs is called opt/venvs - made changes to permissions as above.  

`tensorflow::install_tensorflow()`  works!

**Comment by bogsnork** [comment]
buuut, we're back here: 
```
> model %>% compile(
+   loss = list(class_loss, bbox_loss),
+   optimizer = "adam",
+   metrics = list(
+     class_output = custom_metric("class_loss", metric_fn = class_loss),
+     bbox_output = custom_metric("bbox_loss", metric_fn = bbox_loss))
+ )
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  RuntimeError: Evaluation error: object 'tf' not found. 
```

**Comment by bogsnork** [comment]
Maybe I've had this wrong all along. It's looking for an object called `tf`: 
> `object 'tf' not found.`

`tf` is called in the two loss functions: 
```r
#loss function ----
# shapes are batch_size * 16 * 21
class_loss <- function(y_true, y_pred) {
  class_loss  <-
    tf$nn$sigmoid_cross_entropy_with_logits(labels = y_true, logits = y_pred)
  class_loss <-
    tf$reduce_sum(class_loss) / tf$cast(n_classes + 1, "float32")
  class_loss
}


# shapes are batch_size * 16 * 4
bbox_loss <- function(y_true, y_pred) {
  # calculate localization loss for all boxes where ground truth was assigned some overlap
  # calculate mask
  pos <- y_true[, , 1] + y_true[, , 3] > 0
  pos <- pos %>% k_cast(tf$float32) %>% k_reshape(shape = c(batch_size, 16, 1))
  pos <- tf$tile(pos, multiples = k_constant(c(1L, 1L, 4L), dtype = tf$int32))
  diff <- y_pred - y_true
  # mask out irrelevant activations
  diff <- diff %>% tf$multiply(pos)
  
  loc_loss <- diff %>% tf$abs() %>% tf$reduce_mean()
  loc_loss * 100
}
```

**Comment by bogsnork** [comment]
It isn't strictly an object: it's calling the python library `tf` (i.e. tensorflow) using reticulate grammar:
https://rstudio.github.io/reticulate/#importing-python-modules 
Suggests that for some reason tensorflow hasn't been imported in python, or possibly imported under a different name (i.e. not `tf`).

**Comment by bogsnork** [comment]
Commit 8f6893f seems to have fixed the `tf` error #7 , but: 

```
Error occurred in generator: could not find function "load_and_preprocess_image"
WARNING:tensorflow:Your dataset iterator ran out of data; interrupting training. Make sure that your iterator can generate at least `steps_per_epoch * epochs` batches (in this case, 640 batches). You may need touse the repeat() function when building your dataset.
    
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  ValueError: Empty training data. 
```

**Comment by bogsnork** [comment]
Seems like a bodge but closing this now and returning to main issue for this script #5 .
